### PR TITLE
Feature/issue 412 chi square test

### PR DESCRIPTION
**Pull request by thelseraphim**
#### Summary:

Reworks many of the tests for univariate distributions with some new utility functions and a hopefully much more readable style. Resolves https://github.com/stan-dev/math/issues/412

#### Intended Effect:

Reduces boilerplate code in tests and improves readability.

#### How to Verify:

Run the tests and glance at a few of the tests, which mostly look the same. The action is in `util.hpp` and a typical example usage is in `gamma_test.cpp`.

#### Side Effects:

None

#### Documentation:

More comments than we had before

#### Reviewer Suggestions: 

Do we like this pattern? It could be a lot less repetitive if we had lambdas or had a more OO approach to distributions but I don't want to make that kind of change just for the sake of tests.

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company): Columbia University

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

### CODE REVIEW

**Comment by bob-carpenter** [comment]
Looks like Poisson is failing unit tests:

```
test/unit/math/prim/scal/prob/poisson_test failed
```

File: test/unit/math/prim/scal/prob/util.hpp [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+  std::vector<double> expected;
+	for (int i=0; i<K; i++) {
+	  expected.push_back(expected_count);
+	}
+
+	std::vector<int> counts(K);
+  int current_index = 0;
+	for (int i=0; i<N; ++i) {
+    while (mysamples[i] >= quantiles[current_index]) {
```
Comment (thelseraphim) [inline test/unit/math/prim/scal/prob/util.hpp]: it isn't necessary because I'm sorting the samples first.

**Comment by thelseraphim** [comment]
Made some fixes

- fixed whitespace
- fixed spacing after commas
- removed very obvious comments
- replaced all c-style casts with static_cast
- ran poisson_test locally and it's completely fine

I now realize that make cpplint doesn't cover the unit tests, this sucks! I'll poke around for long lines and fix them now.

File: test/unit/math/prim/scal/prob/util.hpp (hunk 1 above)
Comment (bob-carpenter) [inline test/unit/math/prim/scal/prob/util.hpp]: Nice!  That'll take complexity down from O(N^2) to O(N log N).

**Comment by bob-carpenter** [comment]
You wouldn't want the cpplint output for tests at the
moment!  It'd be an enormous (but perhaps worthwhile)
project to reformat all the tests.  We did it for the
non-test code.   We're not being that picky---I just
wanted to make sure you understood the conventions.

**Comment by bob-carpenter** [comment]
There are still failing tests from what look like includes that Travis picked up.  When the tests pass, this is ready to merge.  Thanks for fixing all the formatting.  This pull request makes me really happy---I love to see cut-and-paste replaced with general calls and I like how this is set up to take a vector of draws.

**Pull request merged**